cmd/kubeadm/app/util/config/initconfiguration.go

```diff
@@ -102,6 +105,11 @@ func SetNodeRegistrationDynamicDefaults(cfg *kubeadmapi.NodeRegistrationOptions,
 		klog.V(1).Infof("detected and using CRI socket: %s", cfg.CRISocket)
 	}
 
+	if isDualStack {
```

Comment:
That won't work since we pass in `cfg *kubeadmapi.NodeRegistrationOptions` and not `ClusterConfiguration`

Reply:
ah i see, so one solution is to make SetNodeRegistrationDynamicDefaults accept ClusterConfiguration (it think it used to be like that at some point) :thinking: 

@rosti WDYT?
/cc @rosti

Reply:
This is not the place to do the feature gate patching for the kubelet. The correct place is `buildKubeletArgMap`:
https://github.com/kubernetes/kubernetes/blob/fc4cbf743ac899df1d96f520951c6b24dcaa8494/cmd/kubeadm/app/phases/kubelet/flags.go#L78

Reply:
@rosti I see a problem with how `BuildArgumentListFromMap` works today -- it does not handle comma separated fields like "feature-gates", since it does a simple overwrite of default values with user specified values. For example, if we set IPv6DualStack=true in base, then that will be overwritten if the user has specified some additional feature-gates in `KubeletExtraArgs`. 

We have a couple of options:
1. Make `BuildArgumentListFromMap` capable of "merging" fields like feature-gates rather than simple key/value overwrite. We would do something like this:
```	
for k, v := range overrideArguments {
		if k == "feature-gates" && len(argsMap[k]) > 0{
			argsMap[k] += "," + v
		}else{
			argsMap[k] = v
		}
	}
```
2. Modify top-level funcs like `WriteKubeletDynamicEnvFile` so that they perform the merge on feature-gates.  This would require updating the `argList` list returned by `BuildArgumentListFromMap`. This is not ideal, IMO.

We could also just leave the code in the higher-level funcs as it is in the current PR. 
I'll wait for you feedback before proceeding. Thanks.

Reply:
Let's keep things simple for now and just allow the users to overwrite the whole of the `feature-gates` flags if they wish.
Either way, this is going to be documented as part of the feature gate documentation.
Another possibility is to modify `BuildArgumentListFromMap` to warn on overwrites.

Certainly, we should not do this in the dynamic defaulting code.

@fabriziopandini @neolit123 WDYT?

Reply:
i think that @rosti is correct here that dynamic defaults is not the best place.

> Let's keep things simple for now and just allow the users to overwrite the whole of the feature-gates flags if they wish.

makes sense to me.

passing the dual stack feature gate to the kubelet will be implicit (if the kubeadm FG is on). so if the user provides their own feature-gates for the kubelet extraArgs, this should overwrite the implicit gate but should also show a warning.

> Either way, this is going to be documented as part of the feature gate documentation.

we might add a sentence about overrides, yes.

Reply:
Done.